src/fc/qemu/hazmat/libceph.py

```diff
@@ -0,0 +1,187 @@
+
+    def close(self):
+        pass
+
+
+class RBD:
+    def list(self, ioctx):
+        return ioctx.rados._rbd("ls", ioctx.name)
+
+    def create(self, ioctx, name, size):
+        ioctx.rados._rbd(
+            "create",
+            f"{ioctx.name}/{name}",
+            "--size",
+            f"{size}B",
+            use_json=False,
+        )
+
+    def remove(self, ioctx, name):
+        ioctx.rados._rbd("rm", f"{ioctx.name}/{name}", use_json=False)
+
+
+class Image:
+    def __init__(self, ioctx, name, snapname=None):
+        self.ioctx = ioctx
+        self.rbd = RBD()
+        self.name = name
+        self.snapname = snapname
+        self.closed = False
+
+        self._name = f"{self.ioctx.name}/{self.name}"
+        if self.snapname:
+            self._name += "{self._name}@{self.snapname}"
+
+        if self.name not in self.rbd.list(self.ioctx):
+            raise ImageNotFound(self.name)
+
+    def _info(self):
+        assert not self.closed
+        return self.ioctx.rados._rbd("info", self._name)
+
+    def size(self):
+        assert not self.closed
+        return self._info()["size"]
+
+    def resize(self, size):
+        assert not self.closed
+        self.ioctx.rados._rbd(
+            "resize", self._name, "--size", f"{size}B", use_json=False
+        )
+
+    def lock_exclusive(self, cookie):
+        assert not self.closed
+        try:
+            self.ioctx.rados._rbd(
+                "lock", "add", self._name, cookie, use_json=False
+            )
+        except Exception:
+            for lock in self.ioctx.rados._rbd("lock", "list", self._name):
+                if lock["id"] == cookie:
+                    # XXX slight issue here - can't identify whether it's an
+                    # exclusive lock, but I'm going to run with it for now.
+                    break
+            else:
+                raise ImageBusy(errno.EBUSY, "Image is busy")
+
+    def list_lockers(self):
+        assert not self.closed
+        # Emulate the librbd format
+        lockers = {"lockers": []}
+        for locker in self.ioctx.rados._rbd("lock", "list", self._name):
+            lockers["lockers"].append(
+                (locker["locker"], locker["id"], locker["address"])
+            )
+        return lockers
+
+    def list_snaps(self):
+        assert not self.closed
+        return self.ioctx.rados._rbd("snap", "list", self._name)
+
+    def create_snap(self, snapname):
+        assert not self.closed
+        self.ioctx.rados._rbd(
+            "snap", "create", f"{self._name}@{snapname}", use_json=False
+        )
+
+    def remove_snap(self, snapname):
+        assert not self.closed
+        self.ioctx.rados._rbd(
+            "snap", "rm", f"{self._name}@{snapname}", use_json=False
+        )
+
+    def unlock(self, cookie):
+        assert not self.closed
+        # This is a tiny bit fishy - because we can't really know whether this
+        # was our lock the whole "locker" handling is ... weird.
```

Comment:
Doesn't this depend on the cookie choice? In some other places we e.g. include the PID (of the caller) in a cookie, and hence are able to differentiate between different process runs.

Reply:
Okay, the PID might have ben set as `locker_id` via the real librbd, but that is not available to us via the CLI use.